### DISCUSSION TITLE
Updating properties, attributes and value with lambdas

### DISCUSSION THREAD

**Discussion by SergeVDev** [jquery/jquery#5602] [Ideas]
Right now, if you want to update a property, attribute or value, you have to break chaining and cache a previous value in a variable.

Overloading `.data()`/`.attr()`/`.prop()`/`.val()` with lambda versions, could help:

```
$('.progress-line').data('progress-value', (v) => v + 10);
$('.text').val((s) => sanitize(s));
```

**Comment by mgol** [comment]
That’s not true; chaining works with setters. It does not work with getters for obvious reasons.

Example: https://jsbin.com/dexatav/1/edit?js

**Reply by SergeVDev** [reply]
Updating a property needs both of them:

```
const prevValue = $('.progress-line').data('progress-value');
$('.progress-line').data('progress-value', prevValue);
```

It could be just one call, and this call could be a chain part:

`$('.progress-line').data('progress-value', (v) => v + 10);`

What if you have several items? You should make the implicit iteration explicit with `.each()`. That is why the current approach breaks chaining.

**Reply by mgol** [reply]
Ah, I see, you meant updating based on the previous value.

This is actually supported for [`.attr()`](https://api.jquery.com/attr/#attr-attributeName-function), [`.prop()`](https://api.jquery.com/prop/#prop-propertyName-function) and [`.val()`](https://api.jquery.com/val/#val-function) but not for [`.data()`](https://api.jquery.com/data/).

For `.data()`, one complexity is that the getter takes `data-*` attributes into account so there’s a question if that logic should fire as well if a function is passed. But it’s worth to think about it.

**Reply by mgol** [reply]
Another piece of complexity is `.data()` allows one to store arbitrary data, including a function. Supporting a similar signature would require wrapping such a function and would thus be a breaking change. On the other hand, adding another signature for this to avoid a breaking change would get us further from what it looks like in `.attr()` or `.val()`. Plus, we’re trying to avoid entirely new APIs unless gains are big and the size increase is small.

**Reply by mgol** [reply]
Also, it’s not just about a breaking change. If we introduced special behavior for function values, we’d break the ability to copy data between elements without extra logic:
```js
elem2.data("key", elem1.data("key"));
```

On the other hand, if your app needs this, you can create a short `.dataBulk(key, fnChangingValue)` jQuery plugin. I don’t see this directly in jQuery, though.

**Reply by SergeVDev** [reply]
>This is actually supported for [.attr()](https://api.jquery.com/attr/#attr-attributeName-function), [.prop()](https://api.jquery.com/prop/#prop-propertyName-function) and [.val()](https://api.jquery.com/val/#val-function) but not for [.data()](https://api.jquery.com/data/).

Oops. I looked for the lambda overloading of `.data()`, found nothing and decided that there are not such overloadings for the rest functions neither. My bad.

**Reply by mgol** [reply]
We discussed this in https://github.com/jquery/jquery/issues/5603, you can find the conclusions (and an example plugin) there.
